File: model/js/RealmModels.js
const AccountSchema = {
  name: 'Account',
  properties: {
    provider: 'string',
    provider_id: { type: 'string',  indexed: true },
    data: { type: 'string',  optional: true },
    tokens: { type: 'list',  objectType: 'Token' },
    user: { type: 'User',  optional: true }
  }
};

const ClusterNodeSchema = {
  name: 'ClusterNode',
  primaryKey: 'id',
  properties: {
    id: 'string',
    files: { type: 'list',  objectType: 'RealmFile' }
  }
};

const MetadataSchema = {
  name: 'Metadata',
  properties: {
    userId: { type: 'string',  indexed: true },
    namespace: 'string',
    key: 'string',
    value: 'string'
  }
};

const PermissionSchema = {
  name: 'Permission',
  properties: {
    user: { type: 'User',  optional: true },
    file: { type: 'RealmFile',  optional: true },
    mayRead: 'bool',
    mayWrite: 'bool',
    mayManage: 'bool'
  }
};

const RealmFileSchema = {
  name: 'RealmFile',
  primaryKey: 'id',
  properties: {
    id: 'string',
    path: 'string',
    owner: { type: 'User',  optional: true },
    cluster_node: { type: 'ClusterNode',  optional: true }
  }
};

const TokenSchema = {
  name: 'Token',
  primaryKey: 'token',
  properties: {
    token: 'string',
    expires: 'date',
    revoked: { type: 'date',  optional: true },
    files: { type: 'list',  objectType: 'RealmFile' },
    account: { type: 'Account',  optional: true },
    app_id: 'string'
  }
};

const UserSchema = {
  name: 'User',
  primaryKey: 'id',
  properties: {
    id: 'string',
    accounts: { type: 'list',  objectType: 'Account' },
    isAdmin: 'bool'
  }
};

module.exports = {
  AccountSchema,
  ClusterNodeSchema,
  MetadataSchema,
  PermissionSchema,
  RealmFileSchema,
  TokenSchema,
  UserSchema
};

File: model/objc/RealmModels.m
#import "RealmModels.h"

@implementation Account

+ (NSArray<NSString *> *)requiredProperties {
    return @[
        @"provider",
        @"provider_id",
    ];
}

+ (NSArray<NSString *> *)indexedProperties {
    return @[
        @"provider_id",
    ];
}

@end


@implementation ClusterNode

+ (NSArray<NSString *> *)requiredProperties {
    return @[
        @"id",
    ];
}

+ (NSString *)primaryKey {
    return @"id";
}

@end


@implementation Metadata

+ (NSArray<NSString *> *)requiredProperties {
    return @[
        @"userId",
        @"namespace",
        @"key",
        @"value",
    ];
}

+ (NSArray<NSString *> *)indexedProperties {
    return @[
        @"userId",
    ];
}

@end


@implementation Permission

@end


@implementation RealmFile

+ (NSArray<NSString *> *)requiredProperties {
    return @[
        @"id",
        @"path",
    ];
}

+ (NSString *)primaryKey {
    return @"id";
}

@end


@implementation Token

+ (NSArray<NSString *> *)requiredProperties {
    return @[
        @"token",
        @"expires",
        @"app_id",
    ];
}

+ (NSString *)primaryKey {
    return @"token";
}

@end


@implementation User

+ (NSArray<NSString *> *)requiredProperties {
    return @[
        @"id",
    ];
}

+ (NSString *)primaryKey {
    return @"id";
}

@end


